test-xk6: flaky failures on Linux/macOS runners due to Go module proxy race for fork commits

# test-xk6: flaky failures on Linux/macOS runners due to Go module proxy race for fork commits

## Summary

The `test-xk6` workflow intermittently fails on Linux and macOS runners when triggered by a pull request from a **fork**. Re-running the exact same failed jobs a few minutes later makes them pass without any code change. The root cause is a timing race between the CI runners and `proxy.golang.org` populating a brand-new pseudo-version.

## Observed behavior

In run [29948532385](https://github.com/grafana/k6/actions/runs/29948532385), triggered by a fork PR (head sha `750f556b`), the matrix produced this split:

| Job | Result |
|-----|--------|
| `test-xk6 (stable, ubuntu-latest)` | ❌ failure |
| `test-xk6 (stable, ubuntu-24.04-arm)` | ❌ failure |
| `test-xk6 (tip, ubuntu-latest)` | ❌ failure |
| `test-xk6 (tip, ubuntu-24.04-arm)` | ❌ failure |
| `test-xk6 (stable, macos-latest)` | ❌ failure |
| `test-xk6 (tip, macos-latest)` | ✅ success |
| `test-xk6 (stable, windows-latest)` | ✅ success |
| `test-xk6 (tip, windows-latest)` | ✅ success |

All failing jobs errored at the `xk6 build` step:

```
INF adding dependency go.k6.io/k6/v2 => github.com/the-onewho-knocks/k6/v2@750f556b9ccb3074959ed788edd1a81d8436fead
INF Cleaning up work directory /tmp/k6foundry...
ERR resolving dependency: executing go command: exit status 1
##[error]Process completed with exit code 1.
```

After re-running the failed jobs ~1 hour later (no code change, same commit), **all jobs passed**. This confirms the failure is not caused by any defect in the PR code.

## Root cause

### How the workflow resolves fork commits

When a PR comes from a fork, [xk6.yml](https://github.com/grafana/k6/blob/master/.github/workflows/xk6.yml) detects that `github.event.pull_request.head.repo.full_name != github.repository` and sets:

```bash
export XK6_K6_REPO="github.com/${PR_REPO}"
```

xk6/k6foundry then adds a Go module replace directive:

```
go.k6.io/k6/v2 => github.com/<fork-owner>/k6/v2 @ <head-sha>
```

Because `GOPRIVATE` is set to `go.k6.io/k6` only — and **not** to `github.com/<fork-owner>/k6` — Go resolves the replacement module through the **public `proxy.golang.org` and `sum.golang.org`** rather than fetching directly from GitHub.

### The race

When a commit is freshly pushed to a fork, `proxy.golang.org` has not yet fetched and published the corresponding pseudo-version. The first request to the proxy triggers a background fetch (cloning the fork, building the module zip, submitting hashes to `sum.golang.org`). This process can take **1–2 minutes** for a large repo like k6.

During that window, concurrent Go client requests for the same pseudo-version receive a transient error from the proxy (exit status 1). The underlying Go error is swallowed by xk6/k6foundry, which surfaces only the generic `ERR resolving dependency: executing go command: exit status 1`.

### Why the platform split is consistent

The failing jobs (Linux x64, Linux arm64, stable macOS) started at **19:03:40Z** and failed fast — all completing by **19:05:27Z**, roughly 1.5 minutes in. This is exactly when `xk6 build` ran the `go` resolution step.

The passing jobs (Windows x64/arm, tip macOS) spent more time on setup (downloading and unzipping the Go tip toolchain, slower provisioning) before reaching the same `go` resolution step — by which point the proxy had finished populating the module.

The proxy's own record confirms the timeline:

```json
{
  "Version": "v2.0.0-20260722185143-750f556b9ccb",
  "Time": "2026-07-22T18:51:43Z"
}
```

The commit was pushed at **18:51:43Z**. The workflow started at **19:03:40Z** — 12 minutes later. The proxy had *started* the fetch but had not yet completed it when the fast Linux runners requested the module.

After the proxy fully populated the entry (visible in `sum.golang.org` having the `h1:` hash), re-running the failed jobs succeeded immediately — with a completely cold module cache, the resolution now works without any retries.

## Impact

This race affects **every fork PR** on every push. It is not limited to this specific PR or contributor. Because the window is short, it manifests as a flaky failure that developers typically work around by re-running CI, without realizing the underlying structural cause.

## Proposed fix

The fix is to make Go resolve the fork module directly from GitHub (bypassing the proxy population race and the checksum database) by extending `GOPRIVATE` to include the fork's GitHub path. In the `Run tests` step of `xk6.yml`, when `XK6_K6_REPO` is set, also export:

```bash
if [ "${{ github.event_name }}" == "pull_request" -a \
     "${PR_REPO}" != "${{ github.repository }}" ]; then
  export XK6_K6_REPO="github.com/${PR_REPO}"
  export GOPRIVATE="go.k6.io/k6,github.com/${PR_REPO}"  # <-- add this
fi
```

With the fork in `GOPRIVATE`, the `go` toolchain fetches it directly via git and skips `proxy.golang.org` and `sum.golang.org` entirely. This eliminates the race: there is no proxy to populate, so the module is always immediately available regardless of how quickly the runner reaches that step.

A secondary improvement would be to add `GONOSUMDB` or `GONOSUMCHECK` coverage for the fork path as a belt-and-suspenders measure, though `GOPRIVATE` already implies both `GONOPROXY` and `GONOSUMDB`.

--- a/.github/workflows/xk6.yml
+++ b/.github/workflows/xk6.yml
@@ -71,9 +71,11 @@ jobs:
           echo "COMMIT_ID=$COMMIT_ID"
           cd .github/workflows/xk6-tests
           go install go.k6.io/xk6/cmd/xk6@master
+          GOPRIVATE="go.k6.io/k6"
           if [ "${{ github.event_name }}" == "pull_request" -a \
                "${PR_REPO}" != "${{ github.repository }}" ]; then
             export XK6_K6_REPO="github.com/${PR_REPO}"
+            GOPRIVATE="${GOPRIVATE},github.com/${PR_REPO}"
           fi
           # The following is a workaround for Windows, cause when using 'shell: bash', the PATH is expressed
           # with ':' as separator, but Go code, running on a Windows OS, expects ';' as separator.
@@ -82,7 +84,7 @@ jobs:
             XPATH="$HOME/sdk/gotip/bin;$XPATH"
           fi
           PATH="$XPATH" \
-          GOPRIVATE="go.k6.io/k6" xk6 build "$COMMIT_ID" \
+          GOPRIVATE="${GOPRIVATE}" xk6 build "$COMMIT_ID" \
             --output ./k6ext \
             --with github.com/grafana/xk6-js-test="$(pwd)/xk6-js-test" \
             --with github.com/grafana/xk6-output-test="$(pwd)/xk6-output-test"

--- a/.github/workflows/xk6.yml
+++ b/.github/workflows/xk6.yml
@@ -75,7 +75,7 @@ jobs:
           if [ "${{ github.event_name }}" == "pull_request" -a \
                "${PR_REPO}" != "${{ github.repository }}" ]; then
             export XK6_K6_REPO="github.com/${PR_REPO}"
-            GOPRIVATE="${GOPRIVATE},github.com/${PR_REPO}"
+            GOPRIVATE="${GOPRIVATE},${XK6_K6_REPO}"
           fi
           # The following is a workaround for Windows, cause when using 'shell: bash', the PATH is expressed
           # with ':' as separator, but Go code, running on a Windows OS, expects ';' as separator.

